### ISSUE TITLE
Questions in running cyber_recorder

### ISSUE THREAD

**Issue by Yimags** [ApolloAuto/apollo#13178]
### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 18.04)**: 18.04
- **Apollo installed from (source or binary)**: 
- **Apollo version (3.5, 5.0, 5.5, 6.0)**: 6.0
- **Output of `apollo.sh config` if on `master` branch**: 

### Steps to reproduce the issue:
- **run simulation together with apollo-6.0 and lgsvl in modular test**
- **run cyber_recorder to record message from channel**
- **fail to play or get information from record**

### Supporting materials (screenshots, command lines, code/script snippets):
![error_recorder](https://user-images.githubusercontent.com/64920943/101993432-fa8c1480-3cba-11eb-8fe4-d032b85fb4f0.png)

**Comment by daohu527** [comment]
Prompt that the file `20201212004326.record` cannot be found, are you sure there is this file in the current directory?

**Comment by Yimags** [comment]
> Prompt that the file `20201212004326.record` cannot be found, are you sure there is this file in the current directory?

The record got from the first command is stored under /Apollo. Is it the right path to save the record? Or I should save it somewhere else?

**Comment by daohu527** [comment]
Yes, if you not specify the file path, the file will be in the current directory, and the default name is "%Y%m%d%H%M%S.record". Since you cannot find the file, you can confirm it in the current directory

**Comment by Yimags** [comment]
> Yes, if you not specify the file path, the file will be in the current directory, and the default name is "%Y%m%d%H%M%S.record". Since you cannot find the file, you can confirm it in the current directory

I can confirm that there is a recording file under /apollo. However instead of xxx.record, it is in the format of xxx.record.00000.
Is that the reason for this problem?
![error_no recording](https://user-images.githubusercontent.com/64920943/102081463-8f545680-3e10-11eb-9ab3-d11cd167e743.png)

**Comment by daohu527** [comment]
Yes, that is the problem

**Comment by Yimags** [comment]
> Yes, that is the problem

Thx for the quick response! Now I could get the info and play the record correctly. However I have still some questions remaining:
1. How could I visualize the process of playing recording? 
2. How could I extract the data from recording I play? 
3. In addition, I noticed the recording file is automatically segmented and given the suffix '.00000' and '.00001' for the following part. So how could I merge them directly when I first make the recording in terminal?

**Comment by daohu527** [comment]
> 1.How could I visualize the process of playing recording?

You can use `cyber_monitor` or `cyber_visualizer` to visualize the message. It will show the channels and topics when playing. [link](https://github.com/ApolloAuto/apollo/blob/1c60ca6058fc6c2c8df52d9c220a2cbe9beb80bc/cyber/doxy-docs/source/CyberRT_Developer_Tools.md)

> 2.How could I extract the data from recording I play?

You can use a node to subscribe the channel you need to extract and write it to files. [link](https://github.com/ApolloAuto/apollo/blob/b0833bc0d93797e91ab9d2971e4843c0b0b4c0e2/cyber/doxy-docs/source/CyberRT_Python_API.md#32-record-interfaces). Here are some examples, but they may not be copied completely


> 3.In addition, I noticed the recording file is automatically segmented and given the suffix '.00000' and '.00001' for the following part. So how could I merge them directly when I first make the recording in terminal?

The suffix is to ensure that the file will not be too large each time, the deafault config is below.
```
  static const uint64_t SEGMENT_INTERVAL_ = 60 * 1000 * 1000 * 1000ULL;  // 60s
  static const uint64_t SEGMENT_RAW_SIZE_ = 2048 * 1024 * 1024ULL;  // 2GB
```
it will split in 60s or 2G file size.

**Comment by Yimags** [comment]
It is quite a detailed guidance!
After I tried and ran it with bazel, it shows like this. Did I wrongly run it or it was designed like this? How should I make it more similar with the textual format like in cyber_monitor?
![error record](https://user-images.githubusercontent.com/64920943/102820533-b3e29c80-43d5-11eb-9397-5cfb5dccbe90.png)

**Comment by daohu527** [comment]
Seems you succeeded run it, the parse msg is you want
```
        print('***After parse(if needed),the message is ->')
        if datatype == MSG_TYPE:
            msg_new = SimpleMessage()
            msg_new.ParseFromString(msg)
            print(msg_new)
        elif datatype == MSG_TYPE_CHATTER:
            msg_new = Chatter()
            msg_new.ParseFromString(msg)
            print(msg_new)
```

**Comment by Yimags** [comment]
```
        print('***After parse(if needed),the message is ->')
        if datatype == MSG_TYPE:
            msg_new = SimpleMessage()
            msg_new.ParseFromString(msg)
            print(msg_new)
        elif datatype == MSG_TYPE_CHATTER:
            msg_new = Chatter()
            msg_new.ParseFromString(msg)
            print(msg_new)
```

Thank you and I have noticed that this section originally exist in the record.py. However, in my case all the parse information isn't shown in terminal. What is the reason for this?
The same thing occcured when I ran listener.py. How could I exactly get those parse messages into terminal and even furthermore store them?
![error listener](https://user-images.githubusercontent.com/64920943/102926645-ac82c800-4495-11eb-8916-e0f97fa6bc7b.png)

**Comment by daohu527** [comment]
## parse
> I tried to change MSG_TYPE = 'apollo.localization.LocalizationEstimate'

Pls check the topic in you record file, if you can find the topic then you can parse it, for example if you want to parse `apollo.localization.LocalizationEstimate`

```
        if datatype == apollo.localization.LocalizationEstimate:
            msg_new = LocalizationEstimate()          # the object define is in the proto file
            msg_new.ParseFromString(msg)
            print(msg_new)
```


## serialize and deserialize 
> record_file.write(msg_new)
The error message has been told what you did wrong. If you want to save an "object", you should serialize it as a string, in protobuf, you can just use
```
desc_str = proto.SerializeToString()
``` 
if you want to read the message in the file, You need to deserialize it.
```
msg_new.ParseFromString(msg)
```

**Comment by Yimags** [comment]
Thanks for the explanation of Parse and serialize! I get still some confusion about implementation.

1. For `msg_new = LocalizationEstimate()          # the object define is in the proto file`
After checking `unit_test.proto`, I found only three types of messages ( `UnitTest`, `Chatter`, `ChatterBanchmark`). So for different message types from CyberRT, am I supposed to add new messages types, for example `LocalizationEstimate() `,for each type in this `unit_test.proto` file? Or they have already been existing somewhere else.

2. For `proto.SerializeToString()`
From with module I could import this? Or it should be customized in other `.proto` file. I simply run this command and it shows a `name 'proto' is not defined` error.

**Comment by daohu527** [comment]
Generally speaking, you need to understand the usage of protobuf, and then deal with the following questions, the official document is better than what I introduced.

1. 
> For msg_new = LocalizationEstimate() # the object define is in the proto file
For different message types from CyberRT, I also found them in .proto file under respective module ducuments. When I want to read only part of the original message, am I supposed to modified the .proto file or just in the python scripts?

You don't need to modify it, because some object is `optional`.

2. 
> For proto.SerializeToString()
From with module I could import this? Or it should be customized in other .proto file. I simply run this command and it shows a name 'proto' is not defined error.

`proto` is the message type, for example you want to serialize `LocalizationEstimate` then you should use 
```
LocalizationEstimate.SerializeToString()
```

**Comment by Yimags** [comment]
By referring to the official documentation and then implement it in code, it seems to me that the output of 
` proto.SerializeToString()` is kind of byte-style string information with structure broken and hard for human reading. I was thinking about storing data just in a stuctured way as it shows with `msg_new.ParseFromString(msg)`. I have found external methods to transfer from protobuf to json and from json to excel. But is there any internal functions to realize it directly?

**Comment by daohu527** [comment]
with c++ you can save with txt format. There is already a interface in Apollo.
https://github.com/ApolloAuto/apollo/blob/a29a563e95944b603ab9be338cce46e6486a89be/cyber/common/file.cc#L56-L64

**Comment by Yimags** [comment]
Thx! I have successfully converted the data to .txt format but not with this interface. I am wondering how I could make use of these c++ functions in python binaries in bazel?

**Comment by daohu527** [comment]
Python already have a interface.

> Finally, each protocol buffer class has methods for writing and reading messages of your chosen type using the protocol buffer binary format. These include:

* SerializeToString(): serializes the message and returns it as a string. Note that the bytes are binary, not text; we only use the str type as a convenient container.
* ParseFromString(data): parses a message from the given string.

[Link](https://developers.google.com/protocol-buffers/docs/pythontutorial#parsing-and-serialization)

**Comment by Yimags** [comment]
Thx for explanation! I exactly am using this python interface in my script as you taught me before.

I got some trivial questions reaḿaining about cyber_recorder.
1. Is it possible to run recorder in python binaries?
2. How could I change the default naming rules?
3. I can't find the script inclues the setings for segementation in `/cyber/record/`. Where did they locate exactly?
```
  static const uint64_t SEGMENT_INTERVAL_ = 60 * 1000 * 1000 * 1000ULL;  // 60s
  static const uint64_t SEGMENT_RAW_SIZE_ = 2048 * 1024 * 1024ULL;  // 2GB
```

**Comment by daohu527** [comment]
> Is it possible to run recorder in python binaries?

What do you mean by `python binaries`, do you mean build it in a binary file and run? If so, yes.

> How could I change the default naming rules?

You can use a `--output` to set the output file's name

> I can't find the script inclues the setings for segementation in /cyber/record/. Where did they locate exactly?

It's in `cyber/record/header_builder.h`

**Comment by Yimags** [comment]
> > Is it possible to run recorder in python binaries?
> 
> What do you mean by `python binaries`, do you mean build it in a binary file and run? If so, yes.
> 

I meant if there is a python interface to run recorder instead of run `cyber_recorder record ` in terminal.

**Comment by daohu527** [comment]
You can try "cyber/python/cyber_py3/record.py"

**Comment by Yimags** [comment]
I have referred to the simple example provided. It shows how to write finite simple messages to certain channel at certain timestep. 
```
    # Writer 2 SimpleMessage
    msg = SimpleMessage()
    msg.text = "AAAAAA"

    file_desc = msg.DESCRIPTOR.file
    proto = FileDescriptorProto()
    file_desc.CopyToProto(proto)
    proto.name = file_desc.name
    desc_str = proto.SerializeToString()
    print(msg.DESCRIPTOR.full_name)
    fwriter.write_channel(
        'simplemsg_channel', msg.DESCRIPTOR.full_name, desc_str)
    fwriter.write_message('simplemsg_channel', msg, 990, False)
    fwriter.write_message('simplemsg_channel', msg.SerializeToString(), 991)
```
But when I want to record an entire channel or more. Do I have to get data from `listener.py` and assign them to `msg` one by one? Or there is a more direct way to realize?

**Comment by daohu527** [comment]
If you want to subscribe to multiple channels, yes, you have to subscribe one by one. This determines your needs.
1. If you want to record data(the data is sorted by time), you can use the tool `cyber_record`
2. If you want to unzip a bag, then use subscribe to multiple channels is enough.

**Comment by Yimags** [comment]
Actually, I have succeed in recording data with cyber_recorder, but I am wondering if I could do it in Python script instead.

**Comment by marcusvinicius178** [comment]
Hi @Yimags I have also recorded the channels with the flag -all and after play this record I was not able to see the graphics on cyber_visualizer:

![image](https://user-images.githubusercontent.com/43826438/123005095-ee7cc000-d38b-11eb-96bb-17dd97a1ad13.png)

I was used with Autoware where I recorded the topics and saved on bags. Afterwards I converted these bags into csv data to further plot these data...

In Apollo, however, I believe to get this data from channels is harder (As I checked in this post). I will need to create a subscriber to get data from a channel and later to convert to .txt and later plot...
I checked the link provided https://github.com/ApolloAuto/apollo/blob/b0833bc0d93797e91ab9d2971e4843c0b0b4c0e2/cyber/doxy-docs/source/CyberRT_Python_API.md#32-record-interfaces
But I could not see a practical example, for example, get data from the /apollo/prediction channel or from the IMU, GPS, Control channels, for example...
**Could you please provide me this piece of code you used to get data from IMU channel, please?I need an example to properly extract data from the channels I need**

And @daohu527 is given a better example of how to get data of these specific channels?

Thanks in advance.

**Comment by marcusvinicius178** [comment]
Hi @Yimags  Thanks very much for the complete answer, I got the idea, however if possible I would like to ask additional questions from my notes. Well let's go for the first option you told me:

1 - Use a Listerner.py: From this option I accessed this link https://github.com/ApolloAuto/apollo/blob/master/modules/tools/plot_control/plot_control.py

Looking at code I have noticed that it plots steering,brake and throttle data "online". I mean while I am running this node and the car is moving the plot is happening. I believe that this is the functionality of those graphics displayed online on Dreamview on PNC Monitor right? And **they are dynamic graphics**. Well I conclude if I wish other data (Odometry, GPS, Control, etc) I need to modify this code and implement my customized Listener.py correct? 

The second option seems a little bit more suitable for me, once I **do not wish "online dynamic graphics",** but to plot all Timestamp data from a window time (from beginning until the end of the simulation). So I would record all the channels from Apollo using the cyber record command. Afterwards as the tutorial you sent me says I can get the info of this recorded channels, and the output will be something similar to this:

![image](https://user-images.githubusercontent.com/43826438/125877959-31ef0ad9-8025-4816-8cc6-7ecd57e4edd2.png)

This is good! Apollo records all channels, all of them contain messages! Awesome. **HOWEVER not all of them are PARSED by the parserrecord parser tool implemented in python**. As far as I have read the standard parsed data **are just related to the following sensors: radar, camera, lidar**. These are the standard sensors d**efined on parser_params.yaml** 

![image](https://user-images.githubusercontent.com/43826438/125878360-12f25e27-8bc5-4c6c-8250-683d900c2489.png)


Therefore after executing the  $ record_parse_save command I will **get just .txt data from these sensors and not from the other channels, and this is bad!** I need the **other channels data converted to .txt** for further plotting this data for my master thesis discussion.

According to the tutorial I must change the code inside parser_params.yaml  if I wish the other data. Is this correct or is some other way to work around?  

If this be the only way than let me give an example to understand if this would work. Let me say I wish to parse data from pose and imu channels, instead those data from camera, radar and lidar . So the block of code instead radar, lidar and camera **would be (on parser_params.yaml) :**

```
pose:     # for position
  channel_name: /apollo/localization/pose
  out_folder_extn: ???
  timestamp_file_extn: pose_timestamp.txt

imu:     # for IMU
  channel_name: /apollo/sensor/gnss/imu
  out_folder_extn: ???
  timestamp_file_extn: imu_timestamp.txt
```

**Do you know to say what exactly The out_folder_extn means and how must I fullfill this area ?** 
Is Just the name of the folder where the data will be saved or extension of something that I don't know what? Extension of sensor data...Well I do not know how to rename this. Some Idea?

Sorry for the bunch of questions, but I really need to plot this data for my master thesis.

Let me know what would you do to get the other channels data parsed on .txt to further plot
and also about the 2 options (listener vs cyber record) would work better and how for all channels data?


Thanks in advance.

**Comment by marcusvinicius178** [comment]
@Yimags I tried the second option at first, and unfortunately the record_parse_save.py did not work. I have reproduced the steps you sent me. But there is something **wrong i**n Apollo code regarding **the path** to localize the file to be parsed.

Please take a look in the error output:

![channels_ok_but_parse_record_not](https://user-images.githubusercontent.com/43826438/126015235-5f6c73d4-ae42-4bbd-97c0-e26e9697ea11.png)


The record file was stored successfully and the info from the messages recorded from multiple channels can also be vizualized above. However the **record_parse_save.py cannot find the parser_params.yaml file,** which exist and is located in the correct folder, as you can check below:

![file_exist](https://user-images.githubusercontent.com/43826438/126015325-536387f8-9c3c-4fef-a9aa-0176224fc0ca.png)

Could you assist me please identify which code is missing and what I should replace to open the folder and find the parser_params.yaml file correct? I think this is the issue right?

**Comment by daohu527** [comment]
1. You should modify the input path in func `read_parameters`. The latest code has fixed this bug!
2. Also if you want to parse lidar you should modify the config in `parser_params.yaml`, seems can't parse all message type once.

**Comment by marcusvinicius178** [comment]
Hi @daohu527 thanks very much for the answer. 


> 1. You should modify the input path in func `read_parameters`. The latest code has fixed this bug!

You were right, part of this error is fixed! I imagine when I install Apollo 6.0 (latest release) following this tutorial: https://www.svlsimulator.com/docs/archive/2020.06/apollo-master-instructions/
It gets an old release.

Now the MOST IMPORTANT! , I have replaced the record_parse_save.py I had for that one you sent me on the provided link above, then I have runned again the command to parse the files, and I got this error: **No module named 'modules drivers'**

![module_error](https://user-images.githubusercontent.com/43826438/126050079-0bd69b36-4e14-4c87-82dc-aeb4940f07d3.png)

Well I believe this parser works for a real test, right? I should have the pyshical camera, radar or lidar? Unfortunately I am working just with the SVL simulator data. I do not have these sensors, I am just executing simulations without a real car.

Or is this error regarding other issue? Maybe this old release repository which I have dowloaded ?


> 2. Also if you want to parse lidar you should modify the config in `parser_params.yaml`, seems can't parse all message type once.

No I don't wish lidar, camera or radar data. I wish to parse other data recorded, **such as Odometer, IMU, control, planning, time.** I

I tried to modify the parser_params.yaml to accept IMU data, as you can check below:

![IMU_parse](https://user-images.githubusercontent.com/43826438/126050900-7568b427-ad30-4105-aabb-ba85233fffb7.png)

 But unfortunately Apollo Team did not create an parse_imu.py module, just the code for camera, lidar and radar:

![Define_sensor_parser](https://user-images.githubusercontent.com/43826438/126050906-a1f5d835-e3e8-434b-b7ba-829e1aa1c8c9.png)

Could you provide me the other sensors/channels parsers please?
Honestly I have no idea how to properly create these parsers for the different channels: odometry, imu, control, planning, etc. 

To complete my master thesis I must to compare 2 different path planners and I **need to plot the channels data I described to compare graphically against each other**. I wish to plot the data from these channels:

![image](https://user-images.githubusercontent.com/43826438/126051012-ef3f5488-98b9-4c5f-9d3a-7e4de0a60109.png)


Considering the listener option and plot functions available on Apollo Dreamview are Dynamic and online plot, I believe the only way to plot these data would be:

1- Record the Channels using cyber recorder
2- Use the parser tool to get this Channel record data parsed  with .txt format 
3- Convert it to CSV or Data frames using python, to plot them more easily.

However as I said I do not know to create a parser for the IMU, Odometry, Control, Planning and the other channels....

**Is it another feasible/possible way to plot this data for further presentation on my master Thesis**?

**I would like to get those plots available on PNC monitor in Dreamview**. They are pretty good and would be enough for me!!! How is it possible?

Some idea @Yimags, @lesun90 ?


Thanks in advance
